test/visual/globalCompositeOperation.js

```diff
@@ -0,0 +1,149 @@
+    height: 300
+};
+
+/**
+ * source https://github.com/mdn/content/blob/main/files/en-us/web/api/canvas_api/tutorial/compositing/example/index.md#utility-functions
+ */
+const colorSphere = (r) => {
+    const c = fabric.util.createCanvasElement()
+    c.width = c.height = r * 2;
+    const ctx = c.getContext('2d');
+    const rotate = (1 / 360) * Math.PI * 2; // per degree
+    const oleft = 0;
+    const otop = 0;
+    for (let n = 0; n < 360; n++) {
+        const gradient = ctx.createLinearGradient(oleft + r, otop, oleft + r, otop + r);
+        gradient.addColorStop(0, "rgba(0,0,0,0)");
+        gradient.addColorStop(0.7, `hsl(${(n + 300) % 360},100%,50%)`);
+        gradient.addColorStop(1, "rgba(255,255,255,1)");
+        ctx.beginPath();
+        ctx.moveTo(oleft + r, otop);
+        ctx.lineTo(oleft + r, otop + r);
+        ctx.lineTo(oleft + r + 6, otop);
+        ctx.fillStyle = gradient;
+        ctx.fill();
+        ctx.translate(oleft + r, otop + r);
+        ctx.rotate(rotate);
+        ctx.translate(-(oleft + r), -(otop + r));
+    }
+    return c;
+};
+
+function createExisting(center = new fabric.Point(size.width,size.height).scalarDivide(2)) {
+    return new fabric.Image(colorSphere(radius * 2), {
+        left: center.x,
+        top: center.y,
+        originX: 'center',
+        originY: 'center'
+    });
+}
+
+function createNew(operation, center = new fabric.Point(size.width, size.height).scalarDivide(2)) {
+    const circles = [
+        {
+            fill: 'rgb(255,0,0)',
+            center: new fabric.Point(100, 200)
+        },
+        {
+            fill: 'rgb(0,0,255)',
+            center: new fabric.Point(220, 200)
+        },
+        {
+            fill: 'rgb(0,255,0)',
+            center: new fabric.Point(160, 100)
+        }
+    ].map(({ fill, center }) => {
+        return new fabric.Circle({
+            radius,
+            globalCompositeOperation: 'lighter',
+            fill,
+            left: center.x,
+            top: center.y,
+            originX: 'center',
+            originY: 'center',
+        });
+    });
+    return new fabric.Group(circles, {
+        left: center.x,
+        top: center.y,
+        originX: 'center',
+        originY: 'center',
+        globalCompositeOperation: operation,
+        objectCaching: true,
+    });
+}
+
+function createPreview(operation) {
+    const r1 = new fabric.Rect({
+        width: 30,
+        height: 30,
+        fill: 'blue',
+        left: 15,
+        top: 15
+    });
+    const r2 = new fabric.Rect({
+        width: 30,
+        height: 30,
+        fill: 'red',
+        globalCompositeOperation: operation
+    });
+    return new fabric.Group([r1, r2]);
+}
+
+QUnit.module('globalCompositeOperation');
+
+OPERATIONS.map(operation => {
+    return {
+        test: operation,
+        code: (canvas, callback) => {
+            canvas.backgroundImage = createExisting();
+            canvas.add(createNew(operation), createPreview(operation));
```

Comment:
this will break as soon as we fix globalCompositeOperation to be group scoped.
We should make the tests already aware of group, so adding all those elements, background, previews and circles in the same group as siblings.

Reply:
please explain

Reply:
global composite operation reacts differently if you are using caching or not.
so if a group is cached a circle inside a group will stop his global composite operation effect on the group itself and won't reach the canvas.
This isn't explained in fabricJS is just a side effect of caching.

global composite operation should introduce a rule like we have for shadow.
If a global composite operation is detected, caching is forced on.
Of course this will bring to some perf issues on edge cases, like a group with shadow and global composite operation.

There is also an added quirk.
I can add global composite operation on groups as a property.
What does it mean? in theory nothing, because it would be then overridden by the value of the objects inside.
But if we have a documented rule, that fabric objects do have global composite operation work from themselves on their immediate container, things becomes predictable for the developer.

Reply:
so the tests are aligned to your comment:
> If a global composite operation is detected, caching is forced on, and the effect of global composite operation is from an object to its container.

Reply:
ok then, at worst some bit will change and we will verify why and if should chnage